File: app/core/security.py
from datetime import datetime, timedelta, timezone
from typing import Any, Optional, Union
import bcrypt
import jwt
from app.core.config import settings
from app.schemas.role import UserRole


def verify_password(plain_password: str, hashed_password: str) -> bool:
    """Verify a plain text password against a stored hash."""
    return bcrypt.checkpw(
        plain_password.encode("utf-8"), hashed_password.encode("utf-8")
    )


def get_password_hash(password: str) -> str:
    """Generate a bcrypt hash of the password."""
    salt = bcrypt.gensalt()
    return bcrypt.hashpw(password.encode("utf-8"), salt).decode("utf-8")


def create_access_token(
    subject: Union[str, Any],
    role: Optional[UserRole] = None,
    expires_delta: timedelta | None = None,
) -> str:
    """Create a JWT access token containing subject and role claims."""
    if expires_delta:
        expire = datetime.now(timezone.utc) + expires_delta
    else:
        expire = datetime.now(timezone.utc) + timedelta(
            minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES
        )

    to_encode = {
        "exp": expire,
        "sub": str(subject),
        "role": role.value if role else UserRole.USER.value,
        "type": "access",
    }
    encoded_jwt = jwt.encode(
        to_encode, settings.SECRET_KEY, algorithm=settings.ALGORITHM
    )
    return encoded_jwt


def create_refresh_token(
    subject: Union[str, Any],
    role: Optional[UserRole] = None,
    expires_delta: timedelta | None = None,
) -> str:
    """Create a JWT refresh token containing subject and role claims."""
    if expires_delta:
        expire = datetime.now(timezone.utc) + expires_delta
    else:
        expire = datetime.now(timezone.utc) + timedelta(
            days=settings.REFRESH_TOKEN_EXPIRE_DAYS
        )

    to_encode = {
        "exp": expire,
        "sub": str(subject),
        "role": role.value if role else UserRole.USER.value,
        "type": "refresh",
    }
    encoded_jwt = jwt.encode(
        to_encode, settings.SECRET_KEY, algorithm=settings.ALGORITHM
    )
    return encoded_jwt

File: src/utils/security.ts
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import { env } from '../config/env';
import { UserRole } from '../models/User';
import { redisService } from '../services/redis.service';

export interface TokenPayload {
  sub: string;
  role: string;
  roles?: string[];
  account_type?: string;
  type: 'access' | 'refresh';
  exp?: number;
  iat?: number;
}

export async function blacklistToken(token: string, expiresInSeconds: number): Promise<void> {
  if (expiresInSeconds > 0) {
    await redisService.set(`blacklist:${token}`, 'true', expiresInSeconds);
  }
}

export async function isTokenBlacklisted(token: string): Promise<boolean> {
  const result = await redisService.get(`blacklist:${token}`);
  return result !== null;
}

export async function verifyPassword(plain: string, hashed: string): Promise<boolean> {
  return bcrypt.compare(plain, hashed);
}

export async function hashPassword(plain: string): Promise<string> {
  const salt = await bcrypt.genSalt(10);
  return bcrypt.hash(plain, salt);
}

export function createAccessToken(
  subject: string,
  role: string = UserRole.USER,
  expiresInMinutes?: number,
  roles?: string[],
  accountType?: string
): string {
  const expiry = (expiresInMinutes ?? env.ACCESS_TOKEN_EXPIRE_MINUTES) * 60;
  const payload: TokenPayload = {
    sub: subject,
    role: role,
    roles: roles,
    account_type: accountType,
    type: 'access',
  };
  return jwt.sign(payload, env.SECRET_KEY, {
    algorithm: 'HS256',
    expiresIn: expiry,
  });
}

export function createRefreshToken(
  subject: string,
  role: string = UserRole.USER,
  expiresInDays?: number,
  roles?: string[],
  accountType?: string
): string {
  const expiry = (expiresInDays ?? env.REFRESH_TOKEN_EXPIRE_DAYS) * 24 * 60 * 60;
  const payload: TokenPayload = {
    sub: subject,
    role: role,
    roles: roles,
    account_type: accountType,
    type: 'refresh',
  };
  return jwt.sign(payload, env.SECRET_KEY, {
    algorithm: 'HS256',
    expiresIn: expiry,
  });
}

export function verifyToken(token: string): TokenPayload {
  return jwt.verify(token, env.SECRET_KEY, { algorithms: ['HS256'] }) as TokenPayload;
}
